pkg/api/parser/json/json.go

```diff
@@ -0,0 +1,72 @@
+package json
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/timescale/promscale/pkg/prompb"
+)
+
+type jsonPayload struct {
+	Labels  map[string]string `json:"labels"`
+	Samples jsonSamples       `json:"samples"`
+}
+
+type jsonSamples []prompb.Sample
+
+func (i *jsonSamples) UnmarshalJSON(data []byte) error {
+	var vv [][]float64
+
+	if err := json.Unmarshal(data, &vv); err != nil {
+		return fmt.Errorf("cannot unmarshal string into array of (timestamp,value) arrays")
+	}
+
+	*i = make([]prompb.Sample, 0, len(vv))
+	for _, tv := range vv {
+		if len(tv) != 2 {
+			return fmt.Errorf("invalid sample, should contain only timestamp and value: %v", tv)
+		}
+
+		t := int64(tv[0])
+
+		// Verify that timestamp doesn't contain fractions so we don't silently
+		// drop precision.
+		if float64(t) != tv[0] {
+			return fmt.Errorf("timestamp cannot contain decimal parts in sample: %v", tv)
+		}
```

Comment:
You cannot do this in validate since you need to unmarshal the data before you are able to check each value.

The docs state the exact specification of this format already,  this nothing new introduced in this PR:
https://github.com/timescale/promscale/blob/master/docs/writing_to_promscale.md#json-streaming-format

Reply:
> unmarshal the data before you are able to check each value.

I meant the validation of a sample.